### PR TITLE
SAMZA-2583: Fix NPE in update SamzaHistogram Gauge values

### PR DESCRIPTION
**Pull request by vishwajith-s**
SamzaHistogram gauges are instantiated via passed in metric registry
newGauge api. The new Gauge api in turn tries to instantiate
HistogramGauge. The newGauge api calls getValue on HistogramGauge
which in turn tries to update the Histogram Gauge values. This means
we are updating Histogram Gauge values in process of creating it.
As a fix, we add  null check in Gauge getValue to avoid NPE.

### CODE REVIEW

File: samza-api/src/main/java/org/apache/samza/metrics/SamzaHistogram.java [hunk 1]
```diff
@@ -81,6 +81,10 @@ public void visit(MetricsVisitor visitor) {
      * and do not have MetricsReporter to update the values.
      */
     public Double getValue() {
+      if (gauges == null) {
```
Comment (prateekm) [inline samza-api/src/main/java/org/apache/samza/metrics/SamzaHistogram.java:84]: Please add an inline comment for when this can happen.

**Comment by prateekm** [comment]
@vishwajith-s Can you please add a unit test for this issue so that you fix doesn't get refactored away. Thanks for the quick fix!

**Comment by vishwajith-s** [comment]
> @vishwajith-s Can you please add a unit test for this issue so that you fix doesn't get refactored away. Thanks for the quick fix!

Done.

**Comment by vishwajith-s** [comment]
> @vishwajith-s Can you please add a unit test for this issue so that you fix doesn't get refactored away. Thanks for the quick fix!

Had to add implement MockMetricRegistry class instead of using Mockito framework as we had to ensure getValue is being called as part of newGauge creation.

**Comment by prateekm** [comment]
@vishwajith-s LGTM, thanks. Can you check why the build failed?

**Comment by mynameborat** [comment]
> > @vishwajith-s Can you please add a unit test for this issue so that you fix doesn't get refactored away. Thanks for the quick fix!
> 
> Had to add implement MockMetricRegistry class instead of using Mockito framework as we had to ensure getValue is being called as part of newGauge creation.

Would stubbing the call to mockMetricsRegistry and respond with an answer that invokes `value.getValue()` work? You can use the invocationContext to get the actual arguments.

**Comment by mynameborat** [comment]
> > > @vishwajith-s Can you please add a unit test for this issue so that you fix doesn't get refactored away. Thanks for the quick fix!
> > 
> > 
> > Had to add implement MockMetricRegistry class instead of using Mockito framework as we had to ensure getValue is being called as part of newGauge creation.
> 
> Would stubbing the call to mockMetricsRegistry and respond with an answer that invokes `value.getValue()` work? You can use the invocationContext to get the actual arguments.

Here is an example - https://github.com/apache/samza/blob/216a87fe4c6fa8e0ed48ad7488e296a88d41376e/samza-core/src/test/java/org/apache/samza/clustermanager/TestContainerPlacementActions.java#L193

**Comment by vishwajith-s** [comment]
> > > @vishwajith-s Can you please add a unit test for this issue so that you fix doesn't get refactored away. Thanks for the quick fix!
> > 
> > 
> > Had to add implement MockMetricRegistry class instead of using Mockito framework as we had to ensure getValue is being called as part of newGauge creation.
> 
> Would stubbing the call to mockMetricsRegistry and respond with an answer that invokes `value.getValue()` work? You can use the invocationContext to get the actual arguments.

Thanks! This simplified the test a lot!

File: samza-api/src/test/java/org/apache/samza/metrics/TestSamzaHistogram.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * under the License.
+ */
+
+package org.apache.samza.metrics;
+
+import org.junit.Test;
+import org.mockito.stubbing.Answer;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
```
Comment (mynameborat) [inline samza-api/src/test/java/org/apache/samza/metrics/TestSamzaHistogram.java:26]: minor: can you put the actual imports instead of the start imports?

**Pull request merged**